### ISSUE TITLE
Keyboard is going crazy

### ISSUE THREAD

**Issue by hype3** [johanmalm/labwc#31]
Hi.

I don't know if this is an issue with LABWC or their is some error in my configuration.
But after a while when I am in a terminal my keyboard is going crazy and backspace starting and cant be stopped (tried both Alacritty & Gnome-terminal).
Move cursor to Firefox and program starts scrolling and can't be stopped and if I move mouse over tabs that are open it starts changing tabs.
So keyboard is going crazy :)

I am swedish speaking and have a 4K monitor connected to Displayport.
Don't know if this is the best solution for wlr-randr. I don't see any other modes than 1024x768 when typing wlr-randr in terminal.

My configs:

- environment
XKB_DEFAULT_LAYOUT=se

- autostart
wlr-randr --output X11-1 --custom-mode 3840x2160@60Hz &
swaybg -c '#113344' >/dev/null 2>&1 &

Using openSUSE Tumbleweed and build LABWC packages using [openSUSE Build Service](https://build.opensuse.org/package/show/home%3Ahype0/labwc)

**Comment by johanmalm** [comment]
Good choice of language sir, same here.

Does the keyboard issue occur when running from X11 and/or TTY.

Not sure what problem is, but if it only happens under X11 it might be related to how we track modifier keys. May need a more sophisticated approach.

**Comment by hype3** [comment]
Yes, I did know that you are Swedish.
Can I choose between Wayland or X11? Or is it Wayland with XWayland if I build with xwayland as option?
I haven't tested with TTY as I use Lightdm as login manager. 
Is my solution with wlr-randr good or is there a better approach or maybe a different program to use.

**Comment by johanmalm** [comment]
I've been developing labwc on a simple arch box with X11 - so haven't been testing running it from X11 much. I guess it's useful for debugging, but it's not really the environment I anticipate running from. It may affect wlr-randr and keyboard handling. Not sure - but will explore.

The easiest way to get to a TTY is probably to press ctrl + alt + F2/F3/F4/etc. If you're able to log in to one of those and run labwc, I'd be interested to know if you still have the keyboard issue. Could you also post the output of wlr-randr from there (preferably as a separate issue). It's strange that you only see 1024x768.

You could try creating a labwc.desktop to run it straight from the display manager:
[https://github.com/johanmalm/labwc/wiki/Display-Manager](https://github.com/johanmalm/labwc/wiki/Display-Manager)

I did run on two screens for a while. I got the resolution from wlr-randr and then ran kanshi with this config file:
 
```
profile {
	output HDMI-A-1 position 1366,0
	output eDP-1 position 0,0
}
```

**Comment by hype3** [comment]
-- X11-1 in autostart and using TTY --
wlr-randr says that I am using DP-1 and all modes are there and no keyboard issue.
I got some error from libseat: [libseat/backend/seatd.c:78] Could not connect to socket /run/seat.d.sock: No such file or directory when I exit from labwc and get back to TTY
also get this in TTY: unkown output X11-1

-- output DP-1 in autostart and using TTY --
I get a black screen nothing works need to reboot

-- output eDP-1 in autostart and using TTY --
wlr-randr says that I am using DP-1 and all modes are there and no keyboard issue.
same error from libseat: [libseat/backend/seatd.c:78] Could not connect to socket /run/seat.d.sock: No such file or directory when I exit from labwc and get back to TTY
also get this in TTY: unkown output eDP-1

-- switched to GDM from lightdm --
using DP-1 or eDP-1 works but I get 1024x768 seems to ignore my config?
tried x11-1 in my config and 3840x2160@60Hz is working but that is the only option that I have in wlr-randr (it was the same with lightdm)
But keyboard is crazy again.

Conclusion:
running from TTY all modes is there and no keyboard issues
Don't know what to put in config to get rid of unkown output error.

Anything else you want me to test?

**Comment by johanmalm** [comment]
Do you get the same keyboard behaviour with sway?

**Comment by hype3** [comment]
I installed Sway and tested and I have no issues with keyboard.

**Comment by hype3** [comment]
With latest labwc 0.3.0 I can't reproduce the bug.
I don't know if it is something that has been fixed in latest wlroots or labwc.

**Comment by johanmalm** [comment]
Thanks for the update. I have looked into it and managed to reproduce it (at least part of what you described) under X11 on Ubuntu 20.04 with wlroots 0.12.0 and labwc 0.1.0. What I experienced was that when a button had been pressed+held and repeat action triggered, releasing the button wouldn't stop the repeat action. The only way to stop the repeat was to press another button. This was not related to modifier keys, but happened when merely pressing keys such as 'a', 'b' and 'c'.

Inserting a printf() at the start of keyboard_key_notify() I discovered that the compositor just didn't seem to see the WL_KEYBOARD_KEY_STATE_RELEASED event for the key that triggered repeat. I also confirmed this using the command below both inside and outside the compositor.

```
sudo libinput debug-events --show-keycodes --verbose
```

Interestingly, I could see the WL_KEYBOARD_KEY_STATE_RELEASED event in a terminal outside the labwc compositor, but not inside it. I'm guessing mutter just doesn't pass on the event to labwc.

I've also tried it with the master branches of wlroots and labwc on openbox + Arch Linux. With this setup, I cannot reproduce the issue.

I suggest we close it for now, and re-open if it can be reproduced with the latest master-branches.
